msprime/simulations.py

```diff
@@ -1528,6 +1528,107 @@ def population_size_trajectory(self, steps):
+            # if f(u) = a exp(bu) then b = log(f(t)/f(s)) / (t-s) for each s < t, so
+            # \int_s^t f(u) du = (a/b) \int_s^t exp(bu) b du = (a/b)(exp(bt) - exp(bs))
+            #    = (t - s) * (f(t) - f(s)) / log(f(t) / f(s))
+            # unless b = 0, of course.
+            assert steps[0] == 0
+            dt = np.diff(steps)
+            dP = np.diff(P)
+            dlogP = np.diff(np.log(P))
+            nz = np.logical_and(dP < 0, P[1:]*P[:-1] > 0)
+            const = (dP == 0)
+            return (np.sum(dt[const] * (P[:-1])[const])
+                    + np.sum(dt[nz] * dP[nz] / dlogP[nz]))
+
+        if steps is None:
+            last_N = max(self.population_size_history[:, self.num_epochs - 1])
+            last_epoch = max(self.epoch_times)
+            steps = sorted(list(set(np.linspace(0, last_epoch + 12 * last_N,
+                                                101)).union(set(self.epoch_times))))
+        if verbose:
+            print("{}\t{}\t{}\t{}\t{}\t{}\t{}\t{}".format(
+                "iter", "mean", "P_diff", "mean_diff", "last_P",
+                "adjust_type", "num_steps", "last_step"))
+        p_diff = m_diff = np.inf
+        last_P = np.inf
+        step_type = "none"
+        n = 0
+        # The factors of 20 here are probably not optimal: clearly, we need to
+        # compute P accurately, but there's no good reason for this stopping rule.
+        # If populations have picewise constant size then we shouldn't need this:
+        # setting steps equal to the epoch boundaries should suffice; while if
+        # there is very fast exponential change in some epochs caution is needed.
+        while n < max_iter and (last_P > rtol or p_diff > rtol/20 or m_diff > rtol/20):
+            last_steps = steps
+            _, P1 = self.coalescence_rate_trajectory(
+                            steps=last_steps, num_samples=num_samples,
+                            min_pop_size=min_pop_size,
+                            double_step_validation=False)
+            m1 = mean_time(last_steps, P1)
+            if last_P > rtol:
+                step_type = "extend"
+                steps = np.concatenate([
+                    steps,
+                    np.linspace(steps[-1], steps[-1] * 1.2, 20)[1:]])
+            else:
+                step_type = "refine"
+                inter = steps[:-1] + np.diff(steps)/2
+                steps = np.concatenate([steps, inter])
+                steps.sort()
+            _, P2 = self.coalescence_rate_trajectory(
+                            steps=steps, num_samples=num_samples,
+                            min_pop_size=min_pop_size,
+                            double_step_validation=False)
+            m2 = mean_time(steps, P2)
+            keep_steps = np.in1d(steps, last_steps)
+            p_diff = max(np.abs(P1 - P2[keep_steps]))
+            m_diff = np.abs(m1 - m2)/m2
+            last_P = P2[-1]
+            n += 1
+            if verbose:
```

Comment:
Not so sure about this ``verbose`` option - this information should be logged really (i.e., logging.debug(...)). Outputting via prints makes it a pain to test. The output is pretty opaque anyway unless you're actually reading the code, so not sure how useful this is in general. I would vote to just comment out these prints to keep them handy for debugging in the future and remove the ``verbose`` option.

Reply:
Hm. I agree, it feels hacky, but we need something to give users a clue what's going on when the automatic discretisation-finder doesn't converge. Pretty much every optimiser or numerical integrator has an option like this, to print out semi-opaque information about what it's doing under the hood. It's messy, but general-purpose numerical integration is unavoidably messy sometimes. Other options:

1. I document the refinement method and what's being printed better. Downside: this will be lengthy.
2. The diagnostic information only gets printed if a "failure to converge" error is thrown.
3. The `verbose` option is changed to `_verbose` and undocumented.

I don't like (2) because it's helpful to look at that output to make sure that nothing wierd is happening even if it thinks it has converged.

About testing: I don't think we need to actually test that it is outputting the correct numbers, since as you say it's useful for debugging and probably needs looking at the code. I have put in a test that *uses* the `verbose` option to make sure it doesn't fail. So, my vote is I guess for a bit more documentation and then leave it alone?

Reply:
I've added a commit using the logger to capture this information; there's an example there as well to show how to use it. There's no need to test the content of the debug output.

If you prefer the existing way, just rebase and drop my commit. Please make sure that the ``verbose`` code path tests don't write out strings to stdout when running ``python3 -m nose -vs``, though.